File: languages/java/src/ads/Solutions.java
package ads;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;

/**
 * Reference algorithm implementations (Java).
 *
 * Behavioral contracts match catalog/problems.json exactly.
 * These are deliberately dependency-free and mirror the C++ / Rust / Python
 * ports 1:1.
 */
public final class Solutions {

    private Solutions() {
    }

    /** Indices of the two values summing to target, or {-1, -1}. */
    public static int[] twoSum(int[] nums, int target) {
        java.util.HashMap<Integer, Integer> seen = new java.util.HashMap<>();
        for (int i = 0; i < nums.length; i++) {
            int complement = target - nums[i];
            Integer j = seen.get(complement);
            if (j != null) {
                int small = Math.min(j, i);
                int large = Math.max(j, i);
                return new int[] {small, large};
            }
            seen.put(nums[i], i);
        }
        return new int[] {-1, -1};
    }

    /** First-occurrence index of target (lower bound), or -1. */
    public static int binarySearch(int[] sorted, int target) {
        int low = 0;
        int high = sorted.length;
        while (low < high) {
            int mid = (low + high) >>> 1;
            if (sorted[mid] < target) {
                low = mid + 1;
            } else {
                high = mid;
            }
        }
        if (low < sorted.length && sorted[low] == target) {
            return low;
        }
        return -1;
    }

    /** New array sorted ascending via merge sort. */
    public static int[] mergeSort(int[] nums) {
        if (nums.length <= 1) {
            return Arrays.copyOf(nums, nums.length);
        }
        int mid = nums.length / 2;
        int[] left = mergeSort(Arrays.copyOfRange(nums, 0, mid));
        int[] right = mergeSort(Arrays.copyOfRange(nums, mid, nums.length));
        return merge(left, right);
    }

    private static int[] merge(int[] left, int[] right) {
        int[] merged = new int[left.length + right.length];
        int i = 0;
        int j = 0;
        int k = 0;
        while (i < left.length && j < right.length) {
            if (left[i] <= right[j]) {
                merged[k++] = left[i++];
            } else {
                merged[k++] = right[j++];
            }
        }
        while (i < left.length) {
            merged[k++] = left[i++];
        }
        while (j < right.length) {
            merged[k++] = right[j++];
        }
        return merged;
    }

    /** New array sorted ascending via in-place quick sort (Lomuto). */
    public static int[] quickSort(int[] nums) {
        int[] values = Arrays.copyOf(nums, nums.length);
        quickSortRange(values, 0, values.length - 1);
        return values;
    }

    private static void quickSortRange(int[] values, int low, int high) {
        if (low >= high) {
            return;
        }
        int pivot = values[high];
        int i = low;
        for (int j = low; j < high; j++) {
            if (values[j] < pivot) {
                swap(values, i++, j);
            }
        }
        swap(values, i, high);
        quickSortRange(values, low, i - 1);
        quickSortRange(values, i + 1, high);
    }

    private static void swap(int[] values, int i, int j) {
        int tmp = values[i];
        values[i] = values[j];
        values[j] = tmp;
    }

    /** Maximum contiguous subarray sum (Kadane); empty -> 0. */
    public static int maxSubarray(int[] nums) {
        if (nums.length == 0) {
            return 0;
        }
        int best = nums[0];
        int running = 0;
        for (int value : nums) {
            running = Math.max(value, running + value);
            best = Math.max(best, running);
        }
        return best;
    }

    /** Longest common subsequence length (full-table DP). */
    public static int lcs(String a, String b) {
        int rows = a.length() + 1;
        int cols = b.length() + 1;
        int[][] table = new int[rows][cols];
        for (int i = 1; i < rows; i++) {
            for (int j = 1; j < cols; j++) {
                if (a.charAt(i - 1) == b.charAt(j - 1)) {
                    table[i][j] = table[i - 1][j - 1] + 1;
                } else {
                    table[i][j] = Math.max(table[i - 1][j], table[i][j - 1]);
                }
            }
        }
        return table[rows - 1][cols - 1];
    }

    /** Maximum value under the 0/1 knapsack constraint (full-table DP). */
    public static int knapsack01(int capacity, int[] weights, int[] values) {
        int[][] dp = new int[weights.length + 1][capacity + 1];
        for (int item = 1; item <= weights.length; item++) {
            int weight = weights[item - 1];
            int value = values[item - 1];
            for (int cap = 0; cap <= capacity; cap++) {
                if (weight <= cap) {
                    dp[item][cap] = Math.max(
                            dp[item - 1][cap], dp[item - 1][cap - weight] + value);
                } else {
                    dp[item][cap] = dp[item - 1][cap];
                }
            }
        }
        return dp[weights.length][capacity];
    }

    /** Levenshtein distance (full-table DP). */
    public static int editDistance(String a, String b) {
        int rows = a.length() + 1;
        int cols = b.length() + 1;
        int[][] table = new int[rows][cols];
        for (int i = 0; i < rows; i++) {
            table[i][0] = i;
        }
        for (int j = 0; j < cols; j++) {
            table[0][j] = j;
        }
        for (int i = 1; i < rows; i++) {
            for (int j = 1; j < cols; j++) {
                int cost = a.charAt(i - 1) == b.charAt(j - 1) ? 0 : 1;
                table[i][j] = Math.min(
                        table[i - 1][j] + 1,
                        Math.min(table[i][j - 1] + 1, table[i - 1][j - 1] + cost));
            }
        }
        return table[rows - 1][cols - 1];
    }

    /** Shortest unweighted path length from start to target, or -1. */
    public static int graphBfs(List<List<Integer>> graph, int start, int target) {
        if (start == target) {
            return 0;
        }
        if (start < 0 || start >= graph.size()) {
            return -1;
        }
        int[] distances = new int[graph.size()];
        Arrays.fill(distances, -1);
        distances[start] = 0;
        java.util.ArrayDeque<Integer> queue = new java.util.ArrayDeque<>();
        queue.add(start);
        while (!queue.isEmpty()) {
            int node = queue.poll();
            for (int neighbor : graph.get(node)) {
                if (distances[neighbor] != -1) {
                    continue;
                }
                distances[neighbor] = distances[node] + 1;
                if (neighbor == target) {
                    return distances[neighbor];
                }
                queue.add(neighbor);
            }
        }
        return -1;
    }

    /** Whether target is reachable from start (iterative DFS). */
    public static boolean graphDfs(List<List<Integer>> graph, int start, int target) {
        if (graph.isEmpty()) {
            return start == target;
        }
        if (start < 0 || start >= graph.size()) {
            return false;
        }
        boolean[] visited = new boolean[graph.size()];
        java.util.ArrayDeque<Integer> stack = new java.util.ArrayDeque<>();
        stack.push(start);
        while (!stack.isEmpty()) {
            int node = stack.pop();
            if (node == target) {
                return true;
            }
            if (visited[node]) {
                continue;
            }
            visited[node] = true;
            List<Integer> neighbors = graph.get(node);
            for (int i = neighbors.size() - 1; i >= 0; i--) {
                int neighbor = neighbors.get(i);
                if (!visited[neighbor]) {
                    stack.push(neighbor);
                }
            }
        }
        return false;
    }

    /** Parse an adjacency-list string like "0:1 2;1:0 2" into a graph. */
    public static List<List<Integer>> parseGraph(String encoded) {
        List<List<Integer>> graph = new ArrayList<>();
        if (encoded.isEmpty()) {
            return graph;
        }
        String[] rows = encoded.split(";");
        for (String row : rows) {
            String[] parts = row.split(":");
            List<Integer> neighbors = new ArrayList<>();
            if (parts.length > 1 && !parts[1].isBlank()) {
                String[] tokens = parts[1].trim().split("\\s+");
                for (String token : tokens) {
                    neighbors.add(Integer.parseInt(token));
                }
            }
            graph.add(neighbors);
        }
        return graph;
    }
}

File: languages/rust/src/solutions.rs
//! Classic algorithm solutions (mirrors the Java/C++/Python tiers).

/// Snippet-sentinel returned by searches that miss.
pub const SENTINEL: i32 = -1;

/// Indices of the two values summing to `target`, or `(-1, -1)`.
pub fn two_sum(nums: &[i32], target: i32) -> (i32, i32) {
    for (i, &left) in nums.iter().enumerate() {
        for (j, &right) in nums.iter().enumerate().skip(i + 1) {
            if left + right == target {
                return (i as i32, j as i32);
            }
        }
    }
    (SENTINEL, SENTINEL)
}

/// First-occurrence index of `target` in a sorted slice (lower bound), or -1.
pub fn binary_search(sorted: &[i32], target: i32) -> i32 {
    let mut low = 0usize;
    let mut high = sorted.len();
    while low < high {
        let mid = low + (high - low) / 2;
        if sorted[mid] < target {
            low = mid + 1;
        } else {
            high = mid;
        }
    }
    if low < sorted.len() && sorted[low] == target {
        low as i32
    } else {
        SENTINEL
    }
}

fn merge(left: &[i32], right: &[i32]) -> Vec<i32> {
    let mut merged = Vec::with_capacity(left.len() + right.len());
    let mut i = 0;
    let mut j = 0;
    while i < left.len() && j < right.len() {
        if left[i] <= right[j] {
            merged.push(left[i]);
            i += 1;
        } else {
            merged.push(right[j]);
            j += 1;
        }
    }
    merged.extend_from_slice(&left[i..]);
    merged.extend_from_slice(&right[j..]);
    merged
}

fn quick_sort_range(values: &mut [i32]) {
    if values.len() <= 1 {
        return;
    }
    let pivot_index = values.len() - 1;
    let pivot = values[pivot_index];
    let mut index = 0usize;
    for j in 0..pivot_index {
        if values[j] < pivot {
            values.swap(index, j);
            index += 1;
        }
    }
    values.swap(index, pivot_index);
    quick_sort_range(&mut values[..index]);
    quick_sort_range(&mut values[index + 1..]);
}

/// New array sorted ascending via merge sort.
pub fn merge_sort(nums: &[i32]) -> Vec<i32> {
    if nums.len() <= 1 {
        return nums.to_vec();
    }
    let mid = nums.len() / 2;
    let left = merge_sort(&nums[..mid]);
    let right = merge_sort(&nums[mid..]);
    merge(&left, &right)
}

/// New array sorted ascending via quick sort.
pub fn quick_sort(nums: &[i32]) -> Vec<i32> {
    let mut values = nums.to_vec();
    quick_sort_range(&mut values);
    values
}

/// Maximum contiguous subarray sum (Kadane); empty -> 0.
pub fn max_subarray(nums: &[i32]) -> i32 {
    if nums.is_empty() {
        return 0;
    }
    let mut best = nums[0];
    let mut running = 0;
    for &value in nums {
        running = value.max(running + value);
        best = best.max(running);
    }
    best
}

/// Longest common subsequence length (full-table DP).
pub fn lcs(a: &str, b: &str) -> i32 {
    let a = a.as_bytes();
    let b = b.as_bytes();
    let rows = a.len() + 1;
    let cols = b.len() + 1;
    let mut table = vec![vec![0; cols]; rows];
    for i in 1..rows {
        for j in 1..cols {
            if a[i - 1] == b[j - 1] {
                table[i][j] = table[i - 1][j - 1] + 1;
            } else {
                table[i][j] = table[i - 1][j].max(table[i][j - 1]);
            }
        }
    }
    table[rows - 1][cols - 1]
}

/// Maximum value under the 0/1 knapsack constraint (full-table DP).
pub fn knapsack_01(capacity: usize, weights: &[i32], values: &[i32]) -> i32 {
    let mut dp = vec![vec![0; capacity + 1]; weights.len() + 1];
    for item in 1..=weights.len() {
        let weight = weights[item - 1] as usize;
        let value = values[item - 1];
        for cap in 0..=capacity {
            if weight <= cap {
                dp[item][cap] = dp[item - 1][cap].max(dp[item - 1][cap - weight] + value);
            } else {
                dp[item][cap] = dp[item - 1][cap];
            }
        }
    }
    dp[weights.len()][capacity]
}

/// Levenshtein distance (full-table DP).
pub fn edit_distance(a: &str, b: &str) -> i32 {
    let a = a.as_bytes();
    let b = b.as_bytes();
    let rows = a.len() + 1;
    let cols = b.len() + 1;
    let mut table = vec![vec![0isize; cols]; rows];
    for i in 0..rows {
        table[i][0] = i as isize;
    }
    for j in 0..cols {
        table[0][j] = j as isize;
    }
    for i in 1..rows {
        for j in 1..cols {
            let cost = if a[i - 1] == b[j - 1] { 0 } else { 1 };
            table[i][j] = (table[i - 1][j] + 1)
                .min(table[i][j - 1] + 1)
                .min(table[i - 1][j - 1] + cost);
        }
    }
    table[rows - 1][cols - 1] as i32
}

/// Adjacency list: rows like `"0:1 2;1:0 2"`.
pub fn parse_graph(encoded: &str) -> Vec<Vec<usize>> {
    if encoded.is_empty() {
        return Vec::new();
    }
    let mut graph = Vec::new();
    for row in encoded.split(';') {
        if let Some((_, neighbors)) = row.split_once(':') {
            let parsed = neighbors
                .split_whitespace()
                .filter_map(|part| part.parse::<usize>().ok())
                .collect();
            graph.push(parsed);
        } else {
            graph.push(Vec::new());
        }
    }
    graph
}

/// Length of the shortest unweighted path between `start` and `target`, or -1.
pub fn graph_bfs(graph: &[Vec<usize>], start: usize, target: usize) -> i32 {
    if start == target {
        return 0;
    }
    if start >= graph.len() {
        return SENTINEL;
    }
    let mut distances = vec![SENTINEL; graph.len()];
    distances[start] = 0;
    let mut queue = std::collections::VecDeque::from([start]);
    while let Some(node) = queue.pop_front() {
        let next_distance = distances[node] + 1;
        for &neighbor in &graph[node] {
            if distances[neighbor] != SENTINEL {
                continue;
            }
            distances[neighbor] = next_distance;
            if neighbor == target {
                return next_distance;
            }
            queue.push_back(neighbor);
        }
    }
    SENTINEL
}

/// Whether `target` is reachable from `start` (recursive DFS).
pub fn graph_dfs(graph: &[Vec<usize>], start: usize, target: usize) -> bool {
    if graph.is_empty() {
        return start == target;
    }
    if start >= graph.len() {
        return false;
    }
    let mut visited = vec![false; graph.len()];

    fn visit(graph: &[Vec<usize>], visited: &mut [bool], node: usize, target: usize) -> bool {
        if node == target {
            return true;
        }
        visited[node] = true;
        for &neighbor in &graph[node] {
            if !visited[neighbor] && visit(graph, visited, neighbor, target) {
                return true;
            }
        }
        false
    }

    visit(graph, &mut visited, start, target)
}